Adds a new section about BBQ to the 'Dense vector' page (#2575)
elastic/docs-content: solutions/search/vector/dense-vector.md

    - Refer to [this overview](../semantic-search.md#using-nlp-models) of the main options
  - You can also [bring your own embeddings](bring-own-vectors.md)
    - Use the `dense_vector` field type
2. Query the index using the [`knn` search](knn.md)

## Better Binary Quantization (BBQ) [bbq]

Better Binary Quantization (BBQ) is an advanced vector quantization method, designed for large-scale similarity search. BBQ is a form of lossy compression for [`dense_vector` fields](https://www.elastic.co/docs/reference/elasticsearch/mapping-reference/dense-vector) that enables efficient storage and retrieval of large numbers of vectors, while keeping results close to those from the original uncompressed vectors.

> BBQ offers significant improvements over scalar quantization by using more sophisticated algorithms to reduce memory usage and computational costs while maintaining high search relevance. BBQ is designed to work in combination with [oversampling](#bbq-oversampling) and reranking, and is compatible with various [vector search algorithms](#bbq-vector-search-algorithms), such as [HNSW](#bbq-hnsw) and [brute force (flat)](#bbq-flat).

Reviewer: ```suggestion
BBQ offers significant improvements over scalar quantization by relying on optimized `bit` level computations to reduce memory usage and computational costs while maintaining high search relevance using pre-computed corrective factors. BBQ is designed to work in combination with [oversampling](#bbq-oversampling) and reranking, and is compatible with various [vector search algorithms](#bbq-vector-search-algorithms), such as [HNSW](#bbq-hnsw) and [brute force (flat)](#bbq-flat).
```

### How BBQ works [bbq-how-it-works]

> BBQ retains the original vector’s dimensionality but transforms the datatype of the dimensions from the original `float32` to `bit`. BBQ also keeps some additional information for each vector that improves the ability to predict the vector similarity of the original vector using the quantized vector. 

Reviewer: ```suggestion
BBQ retains the original vector’s dimensionality but transforms the datatype of the dimensions from the original `float32` to `bit` effectively compressing each vector by 32x plus an additional 14 bytes of corrective data per vector. BBQ uses these pre-computed corrective factors as partial distance calculations to help realize impressively robust approximations of the original vector. 
```

Measuring vector similarity with BBQ vectors requires much less computing effort, allowing more candidates to be considered when using the HNSW algorithm. This often results in better ranking quality and improved relevance compared to the original `float32` vectors.

### Vector search algorithms for BBQ [bbq-vector-search-algorithms]

> BBQ currently supports two vector search algorithms, each suited to different scenarios. You can can configure them by setting the dense vector field’s `index_type`.

Reviewer 2: ```suggestion
BBQ currently supports two vector search algorithms, each suited to different scenarios. You can configure them by setting the dense vector field’s `index_type`.
```

#### `bbq_hnsw` [bbq-hnsw]

> When you set a dense vector field’s `index_type` to `bbq_hnsw`, Elasticsearch uses the HNSW algorithm for fast [kNN search](https://www.elastic.co/docs//solutions/search/vector/knn) on compressed vectors. With the default [oversampling](#bbq-oversampling) applied, it delivers better cost efficiency, lower latency, and improved relevance ranking, making it the best choice for large-scale similarity search.

Reviewer 2: Maybe it's worth being explicit about how to set the "index_type" to "bbq_hnsw". From the example it looks like it's by setting the "type" field within "index_options" to "bbq_hnsw". Probably most users will figure this out but it's a bit confusing that it says you should use the "index_type" when the actual fields are "index_options" and "type"

Reviewer 2: Also you could use the {{es}} attribute for "Elasticsearch" here and throughout the page

Author: Hi @benironside, thanks a lot for your feedback! You’re absolutely right, the wording about the parameters was confusing. I’ve updated it accordingly and also added the {{es}} attributes.

:::{note}

> Starting in version 9.1, `bbq_hnsw` is the default indexing method for new `dense_vector` fields, so you don’t need to specify it explicitly when creating an index. 

Reviewer: ```suggestion
Starting in version 9.1, `bbq_hnsw` is the default indexing method for new `dense_vector` fields with greater than 384 dimensions, so you typically don’t need to specify it explicitly when creating an index.  

Datasets with less than 384 dimensions may see less accuracy and incur a higher overhead cost related to the corrective factors, but we have observed some production datasets perform well even at fairly low dimensions including [tests on e5-small](https://www.elastic.co/search-labs/blog/better-binary-quantization-lucene-elasticsearch).
```

        "dims": 64,
        "index": true,
        "index_options": {
          "type": "bbq_hnsw"
        }
      }
    }
  }
}
```

To change an existing index to use `bbq_hnsw`, update the field mapping:

```console
PUT bbq_hnsw-index/_mapping
{
  "properties": {
    "my_vector": {
      "type": "dense_vector",
      "dims": 64,
      "index": true,
      "index_options": {
        "type": "bbq_hnsw"
      }
    }
  }
}
```

After this change, all newly created segments will use the `bbq_hnsw` algorithm. As you add or update documents, the index will gradually convert to `bbq_hnsw`. 

To apply `bbq_hnsw` to all vectors at once, reindex them into a new index where the `index_options` parameter's `type` is set to `bbq_hnsw`:

:::::{stepper}
::::{step} Create a destination index
```console
PUT my-index-bbq
{
  "mappings": {
    "properties": {
      "my_vector": {
        "type": "dense_vector",
        "dims": 64,
        "index": true,
        "index_options": {
          "type": "bbq_hnsw"
        }
      }
    }
  }
}
```
::::

::::{step} Reindex the data
```console
POST _reindex
{
  "source": { "index": "my-index" }, <1>
  "dest":   { "index": "my-index-bbq" }
}
```

> 1. The existing index that you want to reindex into the newly created index with the `bbq_hnsw` algorithm.

Reviewer: ```suggestion
1. The existing index to be reindexed into the newly created index with the `bbq_hnsw` algorithm.
```

This sentence had an odd flow and I found myself rereading it so suggesting an alt; feel free to ignore

::::

:::::

#### `bbq_flat` [bbq-flat]

When you set a dense vector field’s `index_type` to `bbq_flat`, Elasticsearch uses the BBQ algorithm without HNSW. This option generally requires fewer computing resources and works best when the number of vectors being searched is relatively low.

The following example creates an index with a `dense_vector` field configured to use the `bbq_flat` index type.

```console
PUT bbq_flat-index
{
  "mappings": {
    "properties": {
      "my_vector": {
        "type": "dense_vector",
        "dims": 64,
        "index": true,
        "index_options": {
          "type": "bbq_flat"
        }
      }
    }
  }
}
```

### Oversampling [bbq-oversampling]

> Oversampling is a technique used with BBQ searches to reduce the accuracy loss from compression. Compression lowers the memory footprint by over 95% and improves query latency, but this comes at the cost of some result accuracy. The accuracy loss can be mitigated by oversampling during query time and reranking the top results using the full vector. 

Reviewer 2: ```suggestion
Oversampling is a technique used with BBQ searches to reduce the accuracy loss from compression. Compression lowers the memory footprint by over 95% and improves query latency, at the cost of decreased result accuracy. This decrease can be mitigated by oversampling during query time and reranking the top results using the full vector. 
```
For your consideration
